test/e2e/feature_compatibility_version_upgrade/feature_compatibility_version_upgrade_test.go

```diff
@@ -0,0 +1,58 @@
+package feature_compatibility_version_upgrade
+
+import (
+	"testing"
+	"time"
+
+	mdbv1 "github.com/mongodb/mongodb-kubernetes-operator/pkg/apis/mongodb/v1"
+	e2eutil "github.com/mongodb/mongodb-kubernetes-operator/test/e2e"
+	"github.com/mongodb/mongodb-kubernetes-operator/test/e2e/mongodbtests"
+	f "github.com/operator-framework/operator-sdk/pkg/test"
+)
+
+func TestMain(m *testing.M) {
+	f.MainEntry(m)
+}
+
+func TestFeatureCompatibilityVersion(t *testing.T) {
+	ctx := f.NewContext(t)
+	defer ctx.Cleanup()
+
+	// register our types with the testing framework
+	if err := e2eutil.RegisterTypesWithFramework(&mdbv1.MongoDB{}); err != nil {
+		t.Fatal(err)
+	}
+
+	mdb := e2eutil.NewTestMongoDB()
+	t.Run("Create MongoDB Resource", mongodbtests.CreateMongoDBResource(&mdb, ctx))
+	t.Run("Config Map Was Correctly Created", mongodbtests.AutomationConfigConfigMapExists(&mdb))
+	t.Run("Stateful Set Reaches Ready State", mongodbtests.StatefulSetIsReady(&mdb))
+	t.Run("MongoDB Reaches Running Phase", mongodbtests.MongoDBReachesRunningPhase(&mdb))
+	t.Run("Test Basic Connectivity", mongodbtests.BasicConnectivity(&mdb))
+	t.Run("Test Status Was Updated", mongodbtests.Status(&mdb,
+		mdbv1.MongoDBStatus{
+			MongoURI: mdb.MongoURI(),
+			Phase:    mdbv1.Running,
+		}))
```

Comment:
It could be useful to wrap all of these in a single test function. Looks like we'll end up repeating them quite often.
Maybe something like 

```golang
func BasicFunctionality(mdb *mdbv1.MongoDB, ctx *f.Context) func(*testing.T)  {
	return func(t *testing.T) {
		t.Run("Create MongoDB Resource", mongodbtests.CreateMongoDBResource(mdb, ctx))
		t.Run("Config Map Was Correctly Created", mongodbtests.AutomationConfigConfigMapExists(mdb))
		t.Run("Stateful Set Reaches Ready State", mongodbtests.StatefulSetIsReady(mdb))
		t.Run("MongoDB Reaches Running Phase", mongodbtests.MongoDBReachesRunningPhase(mdb))
		t.Run("Test Basic Connectivity", mongodbtests.BasicConnectivity(mdb))
		t.Run("Test Status Was Updated", mongodbtests.Status(mdb,
			mdbv1.MongoDBStatus{
				MongoURI: mdb.MongoURI(),
				Phase:    mdbv1.Running,
			}))
	}
}
```

What do you think?

Reply:
I think the more we group this functions together it gets more difficult to reason locally about what is going on. As an example we have the `assert` functions that will fail on some location that is not the test in particular, meaning that we always have to resort on parsing the error trace to find where the error originated.

We should definitely find a balance without hiding the actual testing details in too many levels of indirection!

Reply:
I implemented the `BasicFunctionality` in the end.